clippy_lints/src/matches/single_match.rs

```diff
@@ -140,74 +140,128 @@ fn check_opt_like<'a>(
     ty: Ty<'a>,
     els: Option<&Expr<'_>>,
 ) {
-    // list of candidate `Enum`s we know will never get any more members
-    let candidates = &[
-        (&paths::COW, "Borrowed"),
-        (&paths::COW, "Cow::Borrowed"),
-        (&paths::COW, "Cow::Owned"),
-        (&paths::COW, "Owned"),
-        (&paths::OPTION, "None"),
-        (&paths::RESULT, "Err"),
-        (&paths::RESULT, "Ok"),
-    ];
-
-    // We want to suggest to exclude an arm that contains only wildcards or forms the exhaustive
-    // match with the second branch, without enum variants in matches.
-    if !contains_only_wilds(arms[1].pat) && !form_exhaustive_matches(arms[0].pat, arms[1].pat) {
-        return;
+    if check_exhaustive(cx, arms[0].pat, arms[1].pat, ty) {
+        report_single_pattern(cx, ex, arms, expr, els);
     }
+}
 
-    let mut paths_and_types = Vec::new();
-    if !collect_pat_paths(&mut paths_and_types, cx, arms[1].pat, ty) {
-        return;
+/// Resturns true if the given type is one of the standard `Enum`s we know will never get any more
+/// members.
+fn is_known_enum(cx: &LateContext<'_>, ty: Ty<'_>) -> bool {
+    const CANDIDATES: &[&[&str; 3]] = &[&paths::COW, &paths::OPTION, &paths::RESULT];
+    CANDIDATES.iter().any(|&ty_path| match_type(cx, ty, ty_path))
```

Comment:
One more improvement here is to use diagnostic items, which are preferred over paths.

```rust
let Some(adt) = ty.ty_adt_def() else { return false };
matches!(cx.tcx.get_diagnostic_name(adt.did), Some(sym::Cow | sym::Option | sym::Result))
```